### PR TITLE
Fix on error add card activity

### PR DESCRIPTION
**Pull request by Epreuve**
### Summary of changes

 - Ensure that `finish(error)` is always called to avoid causing a hang when adding a card in some cases.

 ### Checklist

 - [x] Added a changelog entry

### CODE REVIEW

File: Drop-In/src/main/java/com/braintreepayments/api/dropin/AddCardActivity.java [hunk 1]
```diff
@@ -374,9 +374,8 @@ public void onError(Exception error) {
             } else if (error instanceof DownForMaintenanceException) {
                 mBraintreeFragment.sendAnalyticsEvent("sdk.exit.server-unavailable");
             }
-
-            finish(error);
         }
+        finish(error);
```
Comment (quinnjn) [inline Drop-In/src/main/java/com/braintreepayments/api/dropin/AddCardActivity.java]: We wouldn't want to finish this activity on any error. Field errors are processed on L361 for example which causes a state change in this activity.

Comment (Epreuve) [reply Drop-In/src/main/java/com/braintreepayments/api/dropin/AddCardActivity.java]: I see now.  This change would effectively change the activity making the feedback on the field annoyingly un actionable.  Let me take another wack at this then.

File: Demo/src/androidTest/java/com/braintreepayments/demo/test/DropInTest.java [hunk 2]
```diff
@@ -121,6 +121,9 @@ public void performsThreeDSecure2FrictionlessVerification() {
         onDevice(withTextStartingWith("created")).check(text(endsWith("authorized")));
     }
 
+    @Ignore("Test fails due to Cardinal Activity changes that resulted in the authorization code " +
```
Comment (quinnjn) [inline Demo/src/androidTest/java/com/braintreepayments/demo/test/DropInTest.java]: Nitpick: I think these tests are unrelated to this PR. We should commit this on master first and then rebase this PR.

**Comment by Epreuve** [comment]
> Why are there no tests for this?

Fixed!  Good call out.

File: Drop-In/src/test/java/com/braintreepayments/api/dropin/view/EditCardViewUnitTest.java [hunk 3]
```diff
@@ -353,6 +354,26 @@ public void setVisibility_toVisibleFocusesCvvWhenOptional() {
...
+    public void isEditCardError_returnsTrue_whenCreditCardError() {
+        ErrorWithResponse error = new ErrorWithResponse(422, stringFromFixture("responses/credit_card_error_response.json"));
+
+        assertTrue(mView.isEditCardError(error));
+    }
+
+    @Test
+    public void isEditCardError_returnsFalse_whenErrorIsNotCreditCardOrUnionPay() {
+        ErrorWithResponse error = new ErrorWithResponse(403, new JSONObject().toString());
+        assertFalse(mView.isEditCardError(error));
```
Comment (sestevens) [inline Drop-In/src/test/java/com/braintreepayments/api/dropin/view/EditCardViewUnitTest.java]:
```suggestion

        assertFalse(mView.isEditCardError(error));
```

File: Drop-In/src/main/java/com/braintreepayments/api/dropin/view/EditCardView.java [hunk 4]
```diff
@@ -114,6 +114,11 @@ public void setMaskCvv(boolean mask) {
         mCardForm.maskCvv(mask);
     }
 
+    public boolean isEditCardError(ErrorWithResponse errors) {
+        return errors.errorFor("unionPayEnrollment") != null ||
```
Comment (hollabaq86) [inline Drop-In/src/main/java/com/braintreepayments/api/dropin/view/EditCardView.java]: why do we need to check for UnionPay specifically here?

File: Drop-In/src/test/java/com/braintreepayments/api/dropin/AddCardActivityUnitTest.java [hunk 5]
```diff
@@ -546,6 +547,21 @@ public void onPaymentMethodNonceCreated_sendsAnalyticsEvent() throws JSONExcepti
         verify(mActivity.braintreeFragment).sendAnalyticsEvent("sdk.exit.success");
     }
 
+    @Test
+    public void nonFormFieldError_callsFinishWithError() {
+        setup(mock(BraintreeFragment.class));
+
+        ErrorWithResponse error = new ErrorWithResponse(422, stringFromFixture("responses/three_d_secure/authentication_response_with_error.json"));
```
Comment (scannillo) [inline Drop-In/src/test/java/com/braintreepayments/api/dropin/AddCardActivityUnitTest.java]: Why is this JSON specific to 3DS authentication?

Comment (crookedneighbor) [reply Drop-In/src/test/java/com/braintreepayments/api/dropin/AddCardActivityUnitTest.java]: If it is specific to 3ds, perhaps the test name should reflect it.

File: Drop-In/src/main/java/com/braintreepayments/api/dropin/view/EditCardView.java (hunk 4 above)
Comment (Epreuve) [reply Drop-In/src/main/java/com/braintreepayments/api/dropin/view/EditCardView.java]: In the case of union pay, there is no `creditCard` key and all errors are nested under the `unionPayEnrollment` key.  This checks both to determine if there are errors in either case.

File: Drop-In/src/test/java/com/braintreepayments/api/dropin/AddCardActivityUnitTest.java (hunk 5 above)
Comment (Epreuve) [reply Drop-In/src/test/java/com/braintreepayments/api/dropin/AddCardActivityUnitTest.java]: The code itself isn't specific to 3D Secure, so the test name reflects that.  However, the case that prompted the bug fix did have to do with 3D Secure, which is the reason for the name of the fixture.

However, the way the code is written, an error could come from a non 3D Secure case and flow through this code path as well, so we did not want to give the impression that _only_ 3DS errors would make it here.

Comment (scannillo) [reply Drop-In/src/test/java/com/braintreepayments/api/dropin/AddCardActivityUnitTest.java]: Curious if the fixture should uphold for a more generic error case as well? Or if we should add another.

Comment (Epreuve) [reply Drop-In/src/test/java/com/braintreepayments/api/dropin/AddCardActivityUnitTest.java]: We opted to rewrite the test to be very specific in testing the error payload, and avoid any unnecessary 3D Secure details in the JSON blob.

Comment (scannillo) [reply Drop-In/src/test/java/com/braintreepayments/api/dropin/AddCardActivityUnitTest.java]: This looks good to me! 👍
